lls/arp.c

```diff
@@ -0,0 +1,541 @@
+		}
+next:
+		index = rte_hash_iterate(arp_cache, &key, &data, &iter);
+	}
+}
+
+int32_t
+arp_cache_get(uint32_t ip, struct arp_cache_entry *entry, uint8_t port)
+{
+	int32_t ret;
+
+	if (arp_cache == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: ARP cache not instantiated, but trying to get an entry\n");
+		return -EINVAL;
+	}
+
+	ret = rte_hash_lookup(arp_cache, &ip);
+	if (ret == -ENOENT) {
+		xmit_arp_req(port, ip, NULL);
+	} else if (ret == -EINVAL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"lss: Invalid params; could not lookup ARP entry\n");
+	} else {
+		struct arp_cache_entry *arp_entry = arp_cache_data[ret];
+		ether_addr_copy(&arp_entry->ha, &entry->ha);
+		ret = 0;
+	}
+	return ret;
+}
+
+int32_t
+arp_cache_del(uint32_t ip)
+{
+	struct arp_cache_req *req;
+	int32_t ret;
+
+	if (arp_cache == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: ARP cache not instantiated, but trying to delete an entry\n");
+		return -EINVAL;
+	}
+
+	ret = rte_mempool_get(arp_cache_req_pool, (void **)&req);
+	if (ret == -ENOENT) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: No memory for request, ARP entry not deleted\n");
+		return ret;
+	}
+	req->ip = ip;
+	req->entry = NULL;
+	rte_ring_enqueue(arp_cache_ring, req);
+	return 0;
+}
+
+int32_t
+arp_cache_add(uint32_t ip, const struct ether_addr *ha, uint8_t port)
+{
+	struct arp_cache_entry *arp_entry;
+	struct arp_cache_req *req;
+	int ret;
+
+	if (arp_cache == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: ARP cache not instantiated, but trying to add an entry\n");
+		return -EINVAL;
+	}
+
+	ret = rte_mempool_get(arp_cache_req_pool, (void **)&req);
+	if (ret == -ENOENT) {
+		RTE_LOG(ERR, MEMPOOL,
+			"No memory for request, ARP entry not added\n");
+		return ret;
+	}
+
+	ret = rte_mempool_get(arp_cache_entry_pool, (void **)&arp_entry);
+	if (ret == -ENOENT) {
+		rte_mempool_put(arp_cache_req_pool, req);
+		RTE_LOG(ERR, MEMPOOL,
+			"No memory for new entry, ARP entry not added\n");
+		return ret;
+	}
+
+	/* We've obtained a new entry and request, now fill them in. */
+	ether_addr_copy(ha, &arp_entry->ha);
+	arp_entry->ts = time(NULL);
+	arp_entry->port = port;
+	arp_entry->stale = 0;
+
+	req->ip = ip;
+	req->entry = arp_entry;
+	rte_ring_enqueue(arp_cache_ring, req);
+	return 0;
+}
+
+/*
+ * Returns 0 if the packet needs to be transmitted,
+ * -1 if it does not need to be transmitted (and can be freed).
+ */
+int
+process_arp(struct gatekeeper_if *iface, struct ether_hdr *eth_hdr,
+	uint16_t off)
+{
+	struct arp_hdr *arp_hdr =
+		(struct arp_hdr *)((char *)(eth_hdr + 1) + off);
+	uint32_t target_ip_be_32;
+
+	/* We only support ARP for resolving IPv4 --> Ethernet addresses. */
+	if (unlikely(arp_hdr->arp_hrd != rte_cpu_to_be_16(ARP_HRD_ETHER) ||
+		     arp_hdr->arp_pro != rte_cpu_to_be_16(ETHER_TYPE_IPv4) ||
+		     arp_hdr->arp_hln != ETHER_ADDR_LEN ||
+		     arp_hdr->arp_pln != sizeof(uint32_t)))
+		return -1;
+
+	/* XXX If sip is not in the same subnet as our IP address, drop. */
```

Comment:
Changed. (Is this as simple as adding a prefix length to the config file and checking against it here?)